astropy/utils/tests/test_data.py

```diff
@@ -372,3 +376,37 @@ def test_is_url_in_cache():
 
     download_file(TESTURL, cache=True, show_progress=False)
     assert is_url_in_cache(TESTURL)
+
+
+@contextlib.contextmanager
+def TemporaryTempDir():
+    """Temporarily redirect all temporary files to a temporary directory"""
+    tempdir = tempfile.mkdtemp()
+    old_tempdir = tempfile.tempdir
+    tempfile.tempdir = tempdir
+    try:
+        yield tempdir
+    finally:
+        tempfile.tempdir = old_tempdir
+        shutil.rmtree(tempdir)
+
+
+def test_get_readable_fileobj_cleans_up_temporary_files():
```

Comment:
You don't even need to do all that.  I should have explained a bit more--in my example I also used the `tmpdir` fixture to the test: https://pytest.org/latest/tmpdir.html

This gives you a temporary directory that's sure to be cleaned up at the end of the test--no try/except or `shutil.rmtree` necessary.

One thing to be careful about with the `tmpdir` object is that it's not just a string, but rather a [`py.path.LocalPath`[(http://pylib.readthedocs.org/en/latest/path.html#py._path.local.LocalPath](http://pylib.readthedocs.org/en/latest/path.html#py._path.local.LocalPath) object, which is an object representing a filesystem path and has several methods defined on it.  That's why in my example I did `tmpdir.listdir()` rather than `os.listdir(tmpdir)`.  

Anyways, the test you wrote is just fine--I'm just pointing out some of the utilities in py.test that make things simpler.  So putting it all together the following test *should* work:

```python
def test_get_readable_fileobj_cleans_up_temporary_files(tmpdir, monkeypatch):
    """checks that get_readable_fileobj leaves no temporary files behind"""
    # Create a 'file://' URL pointing to a path on the local filesystem
    local_filename = get_pkg_data_filename(os.path.join('data', 'local.dat'))
    url = 'file://' + pathname2url(local_filename)

    monkeypatch.setattr(tempfile, 'tempdir', str(tempdir))

    # Call get_readable_fileobj() as a context manager
    with get_readable_fileobj(url) as fileobj:
        pass
    # Get listing of files in temporary directory
    tempdir_listing = tmpdir.listdir()

    # Assert that the temporary file was empty after get_readable_fileobj()
    # context manager finished running
    assert len(tempdir_listing) == 0
```